File: second_pipeline/generate_pooled_sfss.R
library(stringr)
library(vcfR, quietly=TRUE)
library(data.table)

# This function was copied from Thomas Taus' poolSeq R package source code
sample.alleles <- function(p, size, mode=c("coverage", "individuals"), Ncensus=NA, ploidy=2) {
  # determine number of return values
  maxlen <- max(length(p), length(size))
  # check length of a, n and size parameter
  if(maxlen %% length(p) != 0 || maxlen %% length(size) != 0)
    warning("Parameters differ in length and are not multiple of one another: length(p)=", length(p), ", length(size)=", length(size))

  # check mode parameter
  mode <- match.arg(mode)

  # sample allele counts according to the specified method
  return(switch(mode,
                coverage={
                  # if length of 'size' equals '1' then generate target coverage values using the Poisson distribution, otherwise use values of 'size' directly
                  cov <- if(length(size) == 1) rpois(n=maxlen, lambda=size) else size
                  # Change all zeros in 'cov' to ones. This prevents division by zero in the next line, but makes the assumption that all sites have
                  # coverage of at least one. Note that sites with zero coverage are extremely rare, especially for larger values of 'size'.
                  cov[which(cov==0)] <- 1
                  # sample allele frequencies from Binomial distribution based on 'cov' and 'p'
                  p.smpld <- rbinom(n=maxlen, size=cov, prob=p) / cov
                  # return results, including coverage values if they were drawn from a Poisson distribution
                  if(length(size) == 1) data.table(p.smpld=p.smpld, size=cov) else p.smpld
                },
                individuals={
                  # if length of 'size' is larger than 1, then send warning message
                  if(length(size) > 1)
                    warning("Only the first element in 'size' will be used, because sampling mode is 'individuals'")

                  # sample random allele frequencies from Hypergeometric distribution
                  rhyper(nn=maxlen, m=p*Ncensus*ploidy, n=(1-p)*Ncensus*ploidy, k=size[1]*ploidy)/(size[1]*ploidy)
                }))
}

# This function mimics moments' "Spectrum.from_data_dict" function, but with the
# application of pool-seq noise to allele frequencies via the "sample.alleles" function.
# The rounding methods are taken from genomalicious' "dadi_inputs_pools" function
# (Thia, J.A. & Riginos, C. (2019) genomalicious: serving up a smorgasbord of R functions
# for population genomic analyses. bioRxiv. doi: https://doi.org/10.1101/667337).
get_pooled_folded_fs <- function(vcf_name, popinfo, haploid_counts, poolseq_depth, method="counts") {
  num_of_pops <- length(haploid_counts)
  # Import VCF file "vcf_name" as "vcf_table"
  vcf <- read.vcfR(vcf_name, verbose=FALSE)
  vcf_genind <- vcfR2genind(vcf)
  vcf_table <- as.data.frame(vcf_genind@tab)
  # Polarize "vcf_table" to remove repeat columns
  polarized_vcf_table <- vcf_table[, which(lapply(names(vcf_table),
                                                  str_sub, -1, -1) == "0")]
  # Subdivide VCF file by population
  populations <- lapply(unique(popinfo),
                        function(i)
                        { polarized_vcf_table[which(popinfo==i),] })
  allele_counts <- lapply(1:num_of_pops,
                          function(i) {apply(populations[[i]], 2, sum) })
  # Apply noise in order to emulate the effects of pool-seq with the sample.alleles
  # function from Thomas Taus' poolSeq package
  allele_freqs <- lapply(1:num_of_pops,
                         function(i) { allele_counts[[i]] / haploid_counts[i] } )
  pooled_allele_freqs <- lapply(allele_freqs,
                                function(x)
                                {sample.alleles(x,
                                                size=poolseq_depth,
                                                mode="coverage") } )
  # Convert frequencies to counts with one of two different methods.
  if (method == "counts") {
    pooled_allele_counts <- lapply(1:num_of_pops,
                                   function(i)
                                   { lapply(pooled_allele_freqs[[i]]$p.smpld,
                                            function(freq)
                                            { count <- freq * haploid_counts[i]
                                            if (count > 0 && count < 0) 1 else round(count)
                                            }) } )
  } else if (method == "probs") {
    pooled_allele_counts <- lapply(1:num_of_pops,
                                   function(i)
                                   { return(rbinom(n=length(pooled_allele_freqs[[i]]$p.smpld),
                                                   size=haploid_counts[i],
                                                   prob=pooled_allele_freqs[[i]]$p.smpld)) } )
  } else {
    stop("Error: 'method' argument must be either 'counts' or 'probs'.", call.=FALSE)
  }
  # Assemble site frequence spectrum
  fs <- array(0, lapply(haploid_counts, function(i) {i + 1}))
  for (i in 1:length(pooled_allele_counts[[1]])) {
    coord <- sapply(pooled_allele_counts, function(x) { x[[i]] + 1 })
    fs[rbind(coord)] <- fs[rbind(coord)] + 1
    rm(coord)
    gc()
  }

  return(fs)
}

generate_pooled_sfs_from_pipeline_instruction <-
  function(vcf_file, popinfo_eval_string, haploid_counts_eval_string, poolseq_depth,
           iterations, wd_string,
           seed) {
  message(seed)
  set.seed(seed)
  fs <- get_pooled_folded_fs(paste(wd_string, "vcfs/", vcf_file, sep=""),
                             eval(parse(text=popinfo_eval_string)),
                             eval(parse(text=haploid_counts_eval_string)),
                             poolseq_depth)
  output_file_name <- paste(wd_string, "serialized_pooled_sfss/",
                            str_sub(vcf_file, end=-5),
                            "_depth", poolseq_depth,
                            "_seed", seed,
                            "_pooled_sfs_serialized.txt", sep="")
  write(c(dim(fs), "-----", rev(fs)), output_file_name, ncolumns=1)
}

# REQUIRES SUBDIRECTORY "serialized_pooled_sfss/" IN WORKING DIRECTORY

# Operational variables
wd_string <- "/scratch/djb3ve/data/second_pipeline/"
setwd(wd_string)
iterations <- 3

# Handle command line arguments
args <- commandArgs(trailingOnly=TRUE)
vcf_file <- args[1]
popinfo_eval_string <- args[2]
haploid_counts_eval_string <- args[3]
poolseq_depth <- as.numeric(args[4])

lapply(1:iterations, function(seed)
  {generate_pooled_sfs_from_pipeline_instruction(vcf_file, popinfo_eval_string,
                                                 haploid_counts_eval_string, poolseq_depth,
                                                 iterations, wd_string, seed)})

File: second_pipeline/generate_pooled_sfss.py
import sys
import os
import numpy as np
import pandas as pd

# Returns number of non-reference alleles in list argument "row", whose elements are
# pipe-separated VCF allele records.
# Treats all sites as biallelic, i.e. all alternate alleles are treated as a single
# "non-reference" allele.
def get_allele_count_from_vcf_row(row):
    cumulative_count = 0
    for polymorphism in row:
        allele_list = polymorphism.split("|")
        alt_allele_count = len(allele_list) - allele_list.count("0")
        cumulative_count += alt_allele_count
    return cumulative_count

# Replicate of "sample.alleles(mode="coverage")" in Thomas Taus' poolSeq
# (https://github.com/ThomasTaus/poolSeq/blob/master/R/simaf.R).
# Lacks support for specifying pool-seq coverage at each site with vector argument for "size" parameter.
# Lacks support for mode="individuals".
def get_pooled_allele_freqs(allele_freqs, poolseq_depth, ploidy=2):
    cov = np.random.poisson(lam=poolseq_depth, size=len(allele_freqs))
    cov = np.where(cov == 0, 1, cov)
    p_sampled = np.random.binomial(n=cov, p=allele_freqs, size=len(allele_freqs)) / cov
    return p_sampled

# Replicate of moments' "Spectrum.from_data_dict" function, but with the
# application of pool-seq noise to allele frequencies
# (https://moments.readthedocs.io/en/devel/api/api_moments.html#moments.Spectrum.from_data_dict).
# The rounding methods are taken from genomalicious' "dadi_inputs_pools" function
# (https://rdrr.io/github/j-a-thia/genomalicious/man/dadi_inputs_pools.html).
def get_pooled_folded_fs(vcf_file, popinfo_file, haploid_counts, poolseq_depth, rounding_method="counts", ploidy=2):
    # Load in files as pandas data frames
    vcf_df = pd.read_table(vcf_file, header=5)
    # Remove non-biallelic polymorphisms
    vcf_df = vcf_df.loc[vcf_df["ALT"].isin(["A", "T", "G", "C"])]
    popinfo_df = pd.read_table(popinfo_file, header=None)

    # Number of rows of "vcf_df" is the number of polymorphic sites recorded in the VCF
    num_of_polymorphisms = vcf_df.shape[0]

    # Subdivide columns of VCF file by population, with one population per element of
    # "vcf_dfs_by_population"
    vcf_dfs_by_population = []
    for pop_id in popinfo_df[1].unique():
        tsks = popinfo_df[popinfo_df[1] == pop_id][0]
        vcf_dfs_by_population.append(vcf_df[tsks])

    allele_freqs_by_population = []
    for i, pop_vcf_df in enumerate(vcf_dfs_by_population):
        allele_counts = pop_vcf_df.apply(get_allele_count_from_vcf_row, axis=1)
        allele_freqs_by_population.append(allele_counts / haploid_counts[i])

    poolseq_allele_freqs_and_coverage_by_population = []
    for pop_allele_freqs in allele_freqs_by_population:
        poolseq_allele_freqs_and_coverage = get_pooled_allele_freqs(pop_allele_freqs, poolseq_depth, ploidy)
        poolseq_allele_freqs_and_coverage_by_population.append(poolseq_allele_freqs_and_coverage)

    pooled_allele_counts_by_population = []
    # The "counts" method intuitively multiplies pooled allele frequencies by haploid counts
    # and then rounds to the nearest integer to yield allele counts. In order to avoid
    # erasing rare alleles, pre-rounding allele counts less than one are always rounded up.
    if rounding_method == "counts":
        for i, pop_pooled_allele_freqs_and_coverage in enumerate(poolseq_allele_freqs_and_coverage_by_population):
            # Scale frequences by haploid counts to achieve allele counts once rounding has been performed
            pop_pooled_allele_counts = pop_pooled_allele_freqs_and_coverage * haploid_counts[i]
            # Prevent rare alleles from getting rounded down to zero, and thus removed from the model
            for j, count in enumerate(pop_pooled_allele_counts):
                if 0 < count < 0.5:
                    pop_pooled_allele_counts[j] = 1
            # Round scaled allele frequencies to achieve allele counts
            pop_pooled_allele_counts = np.round(pop_pooled_allele_counts)
            pooled_allele_counts_by_population.append(pop_pooled_allele_counts)
    # The "probs" method uses binomial sampling in order to avoid the issue of having to round
    # rare alleles counts up in the "counts" method. This introduces more noise and
    # is promising, but should probably not be used as a default.
    # THIS HAS NOT BEEN TESTED YET.
    elif rounding_method == "probs":
        for i, pop_pooled_allele_freqs_and_coverage in enumerate(poolseq_allele_freqs_and_coverage_by_population):
            pop_pooled_allele_counts = np.random.binomial(n=haploid_counts[i],
                                                          p=pop_pooled_allele_freqs_and_coverage,
                                                          size=num_of_polymorphisms)
            pooled_allele_counts_by_population.append(pop_pooled_allele_counts)
    else:
        print("Error: Invalid rounding method.")
        sys.exit()

    # Construct numpy array containing SFS from pooled_allele_counts_by_population
    incremented_haploid_counts = [i + 1 for i in haploid_counts]
    sfs = np.zeros(incremented_haploid_counts)
    for i in range(num_of_polymorphisms):
        coords = []
        for pop_pooled_allele_counts in pooled_allele_counts_by_population:
            coords.append(int(pop_pooled_allele_counts[i]))
        sfs[tuple(coords)] += 1

    return sfs

def main(vcf_file, popinfo_file, haploid_counts, poolseq_depth, output_file_without_suffix, num_of_replicates, rounding_method="counts", ploidy=2):
    # Iterate once per replicate, incrementing the seed between replicates
    for seed in range(1, num_of_replicates+1):
        np.random.seed(seed)
        # Get and serialize SFS
        sfs = get_pooled_folded_fs(vcf_file, popinfo_file, haploid_counts, poolseq_depth, rounding_method, ploidy)
        np.save(f"{output_file_without_suffix}_seed{seed}.npy", sfs)

# Prevent numpy from using scientific notation
np.set_printoptions(suppress=True)

# Set up working environment
working_dir = "/scratch/djb3ve/Demographic-inference-with-pool-seq-data/data/second_pipeline/"
for subdir in ["vcfs/", "serialized_pooled_sfss/"]:
    if not os.path.exists(working_dir + subdir):
        os.makedirs(working_dir + subdir)

# Handle command line arguments, which will be input from "pipeline_instructions.txt",
# except for "num_of_replicates", which is added to the "$OPTS" variable in the wrapper
# script.
vcf_file = working_dir + "vcfs/" + sys.argv[1]
popinfo_file = working_dir + "popinfos/" + sys.argv[2]
output_file_without_suffix = working_dir + "sfss/" + sys.argv[1][:-4] + "_depth" + sys.argv[4]
haploid_counts = eval(sys.argv[3])
poolseq_depth = int(sys.argv[4])
rounding_method = sys.argv[5]
ploidy = int(sys.argv[6])
num_of_replicates = int(sys.argv[7])


main(vcf_file, popinfo_file, haploid_counts, poolseq_depth, output_file_without_suffix, num_of_replicates, rounding_method, ploidy)
